OptimizelySDK.Tests/OdpTests/OdpEventManagerTests.cs

```diff
@@ -0,0 +1,478 @@
+
+            eventManager.Start();
+            eventManager.SendEvent(_testEvents[0]);
+            cde.Wait();
+
+            var eventsSentToApi = eventsCollector.FirstOrDefault();
+            var actualEvent = eventsSentToApi?.FirstOrDefault();
+            Assert.IsNotNull(actualEvent);
+            Assert.AreEqual(expectedEvent.Type, actualEvent.Type);
+            Assert.AreEqual(expectedEvent.Action, actualEvent.Action);
+            Assert.AreEqual(expectedEvent.Identifiers, actualEvent.Identifiers);
+            Assert.AreEqual(expectedEvent.Data["key-1"], actualEvent.Data["key-1"]);
+            Assert.AreEqual(expectedEvent.Data["key-2"], actualEvent.Data["key-2"]);
+            Assert.AreEqual(expectedEvent.Data["key-3"], actualEvent.Data["key-3"]);
+            Assert.AreEqual(expectedEvent.Data["key-4"], actualEvent.Data["key-4"]);
+            Assert.AreEqual(expectedEvent.Data["idempotence_id"].ToString().Length,
+                actualEvent.Data["idempotence_id"].ToString().Length);
+            Assert.AreEqual(expectedEvent.Data["data_source"], actualEvent.Data["data_source"]);
+            Assert.AreEqual(expectedEvent.Data["data_source_type"],
+                actualEvent.Data["data_source_type"]);
+            Assert.AreEqual(expectedEvent.Data["data_source_version"],
+                actualEvent.Data["data_source_version"]);
+        }
+
+        [Test]
+        public void ShouldAttemptToFlushAnEmptyQueueAtFlushInterval()
+        {
+            var eventManager =
+                new OdpEventManager(_odpConfig, _mockApiManager.Object, _mockLogger.Object, 10, 10,
+                    100);
+
+            eventManager.Start();
+            // do not add events to the queue, but allow for...
+            Task.Delay(400)
+                .Wait(); // at least 3 flush intervals executions (giving a little longer)
+
+            _mockLogger.Verify(l => l.Log(LogLevel.DEBUG, "Processing Queue (flush)"),
+                Times.AtLeast(3));
+        }
+
+        [Test]
+        public void ShouldDispatchEventsInCorrectNumberOfBatches()
+        {
+            _mockApiManager.Setup(a =>
+                    a.SendEvents(It.IsAny<string>(), It.IsAny<string>(),
+                        It.IsAny<List<OdpEvent>>()))
+                .Returns(false);
+            var eventManager =
+                new OdpEventManager(_odpConfig, _mockApiManager.Object, _mockLogger.Object, 10, 10,
+                    500);
+
+            eventManager.Start();
+            for (int i = 0; i < 25; i++)
+            {
+                eventManager.SendEvent(MakeEvent(i));
+            }
+            Task.Delay(1000).Wait();
+
+            // Batch #1 & #2 with 10 in each should send immediately then...
+            // Batch #3 of 5 should send after flush interval
+            _mockApiManager.Verify(a =>
+                a.SendEvents(It.IsAny<string>(), It.IsAny<string>(),
+                    It.IsAny<List<OdpEvent>>()), Times.Exactly(3));
```

Comment:
I think `_mockApiManager` is seeing events from other tests because of the `SetNewTimeout()` from the previous (even serial) unit test methods are still executing on the thread pool.

Assuming this unit test suite stays serial, I almost think that each test method should have `Thread.Sleep()` to ensure the previous thread has completed.  

...acutally, perhaps each unit test should have a mock of their own upon which to assert. 🤔 ...instead of sharing the `_mock*`

@msohailhussain